### ISSUE TITLE
Info on Concurrency

### ISSUE THREAD

**Issue by fny** [msiemens/tinydb#487]
It would be great to have info on whether tiny supports concurrent reads and writes. This is not clear from the README.

**Comment by p-baum** [comment]
I would like to know as well please.

**Comment by VermiIIi0n** [comment]
> It would be great to have info on whether tiny supports concurrent reads and writes. This is not clear from the README.

Unfortunately, I believe it's not possible after viewing the source code. 

Concurrent writing/reading almost certainly leads to data corruption.

**Comment by fny** [comment]
Yeah, its a shame.

**Comment by FeralRobot** [comment]
documentation update needed:
class tinydb.middlewares.ConcurrencyMiddleware(storage_cls)
Makes TinyDB working with multithreading.
Uses a lock so write/read operations are virtually atomic.

**Comment by andryyy** [comment]
I am using TinyDB in a project right now and came across this problem.

Since I'm already using Redis for synchronizing some application states in my cluster, I found it easiest to reuse it for distributed locking.

My application uses a Storage class that's pretty much a copy of JSONStorage with some personal tweaks like automatic backups and custom caching.

In the custom storage's `__init__()` definition I added this (slightly modified):

```
...
        self._access_id = kwargs.pop("access_id")
        try:
            if r.get("DATABASE_LOCK") != self._access_id:
                while not r.set(
                    "DATABASE_LOCK",
                    self._access_id,
                    px=4000, # Max lock time in ms
                    nx=True,
                ):
                    continue
        except Exception as e:
            r.delete("DATABASE_LOCK")
            raise
...
```

The `close` method looks like this:

```
...
        self._handle.close()
        r.delete("DATABASE_LOCK")
...
```

TinyDB is used like this:

```
TINYDB = {
    "storage": JSONStorageLocked,
    "path": "database/data.json",
    "access_id": str(uuid4()),
}
with TinyDB(**defaults.TINYDB) as db:
    ...
```

Since the storage instance is reinitialized during operations, I was not able to use id(self) but had to define a fixed ID to use inside the whole context.

At first I tried to only hold a lock inside the `read` and `write` definitions, but in stress tests it of course failed sometimes. The application would lock for a data read, unlock, lock for a write and append the data. But between read and write there is a tiny window for another worker to lock and write data, so I dropped that idea.

Of course I could switch back to that logic and check for the last writer's ID, but I'm fine with it now. :) I also use some caching in Redis, so it does not really impact the performance anyway for the few writes I'm doing.
